chainstate/tx-verifier/src/transaction_verifier/reward_distribution.rs

```diff
@@ -0,0 +1,637 @@
+    accounting_adapter: &mut PoSAccountingDeltaAdapter<P>,
+    block_id: Id<Block>,
+    pool_id: PoolId,
+    total_reward: Amount,
+) -> Result<AccountingBlockRewardUndo, ConnectTransactionError> {
+    let pool_data = accounting_adapter
+        .accounting_delta()
+        .get_pool_data(pool_id)?
+        .ok_or(ConnectTransactionError::PoolDataNotFound(pool_id))?;
+
+    let pool_owner_reward = calculate_pool_owner_reward(total_reward, &pool_data).ok_or(
+        ConnectTransactionError::PoolOwnerRewardCalculationFailed(block_id, pool_id),
+    )?;
+
+    let total_delegations_reward = (total_reward - pool_owner_reward).ok_or(
+        ConnectTransactionError::PoolOwnerRewardCannotExceedTotalReward(
+            block_id,
+            pool_id,
+            pool_owner_reward,
+            total_reward,
+        ),
+    )?;
+
+    // Distribute reward among delegators.
+    // In some cases this process can yield reward unallocated to delegators. This reward goes to the pool owner.
+    let (delegation_undos, unallocated_reward) = if total_delegations_reward > Amount::ZERO {
+        match accounting_adapter.accounting_delta().get_pool_delegations_shares(pool_id)? {
+            Some(delegation_shares) => {
+                let total_delegations_balance =
+                    delegation_shares.values().try_fold(Amount::ZERO, |acc, v| {
+                        (acc + *v).ok_or(ConnectTransactionError::DelegationsRewardSumFailed(
+                            block_id, pool_id,
+                        ))
+                    })?;
+
+                if total_delegations_balance > Amount::ZERO {
+                    let (delegation_undos, delegations_reward_remainder) =
+                        distribute_delegations_pos_reward(
+                            accounting_adapter,
+                            &delegation_shares,
+                            block_id,
+                            pool_id,
+                            total_delegations_balance,
+                            total_delegations_reward,
+                        )?;
+                    (delegation_undos, delegations_reward_remainder)
+                } else {
+                    // If total balance of all delegations is 0 then give the reward to the pool's owner
+                    (Vec::new(), total_delegations_reward)
+                }
+            }
+            // If no delegations then give the reward to the pool's owner
+            None => (Vec::new(), total_delegations_reward),
+        }
+    } else {
+        // Do nothing if no delegations
+        (Vec::new(), Amount::ZERO)
+    };
+
+    let total_owner_reward = (pool_owner_reward + unallocated_reward)
+        .ok_or(ConnectTransactionError::RewardAdditionError(block_id))?;
+    let increase_pool_balance_undo = accounting_adapter
+        .operations(TransactionSource::Chain(block_id))
+        .increase_pool_pledge_amount(pool_id, total_owner_reward)?;
+
```

Comment:
I guess at this point we can add a check, something like pool_owner_reward + unallocated_reward + allocated reward = total reward.

I see it's challenging because we may need to keep track of the distributed rewards... if you can easily do that, please go for it. Otherwise, skip it.

Basically I'm trying to enforce as many checks as humanly possible, with tons of tests ensuring that nothing can go in this distribution. The last thing we want is someone finding a way to print money by staking.

Reply:
you are right it's not that straightforward. Suggest skipping